rbac_api_with_domains.go

```diff
@@ -60,55 +60,70 @@ func (e *Enforcer) DeleteRolesForUserInDomain(user string, domain string) (bool,
 }
 
 // GetAllUsersByDomain would get all users associated with the domain.
+// index denotes the index of domain in p section starting at 0.
 func (e *Enforcer) GetAllUsersByDomain(domain string) []string {
 	m := make(map[string]struct{})
 	g := e.model["g"]["g"]
-	if len(g.Tokens) != 3 {
-		return []string{}
-	}
+	p := e.model["p"]["p"]
 	users := make([]string, 0)
-	for _, policy := range g.Policy {
-		if _, ok := m[policy[2]]; policy[2] == domain && ok {
-			users = append(users, policy[0])
+	index := e.getDomainIndex("p")
+
+	getUser := func(index int, policies [][]string) {
+		if len(policies) == 0 || len(policies[0]) <= index {
+			return
+		}
+		for _, policy := range policies {
+			if _, ok := m[policy[0]]; policy[index] == domain && !ok {
+				users = append(users, policy[0])
+				m[policy[0]] = struct{}{}
+			}
 		}
 	}
+	getUser(2, g.Policy)
+	getUser(index, p.Policy)
 	return users
 }
 
 // DeleteAllUsersByDomain would delete all users associated with the domain.
 func (e *Enforcer) DeleteAllUsersByDomain(domain string) (bool, error) {
 	g := e.model["g"]["g"]
-	if len(g.Tokens) != 3 {
-		return false, nil
-	}
-	policies := make([][]string, 0)
-	for _, policy := range g.Policy {
-		if policy[3] == domain {
-			policies = append(policies, policy)
+	p := e.model["p"]["p"]
+	users := make([][]string, 0)
+	index := e.getDomainIndex("p")
+
+	getUser := func(index int, policies [][]string) {
```

Comment:
why not return users in func? I suggest removing all implicit args like domain, users